File: functions/registrationLimits.js
// ══════════════════════════════════════════════════════════════════════════
// 등록 개수 제한 — **유형별 각각 10개**의 정본.
//
// 한 사업자 계정이 동시에 운영할 수 있는 등록물 수를 유형마다 따로 센다.
// 합산이 아니다: 파티 10 + 이벤트 10 + 플레이스 10 + 장소대여 10까지 된다.
//
// ── 이름 주의 ────────────────────────────────────────────────────────────
// 이 저장소의 컬렉션 이름과 사람이 부르는 이름이 어긋나 있다. 여기서는
// contentCleanup.js의 어휘를 **그대로** 따른다(정리 로직과 짝이 맞아야 한다).
//
//   키          컬렉션              사람이 부르는 이름
//   party       parties             파티
//   promotion   placePromotions     이벤트 (매장 이벤트 · 공간 이벤트)
//   event       events              플레이스 (술집·바·카페)
//   place       places              장소대여 (공간대여·숙박)
//
// ── 무엇이 자리를 차지하는가 ─────────────────────────────────────────────
// **지금 살아 있는 것만** 센다 — 손님에게 보이거나, 곧 보일 것들이다.
//
//   파티        최종 종료 시각이 아직 안 지난 것(무기한 정기 파티 포함)
//   이벤트      진행 중 · 시작 전 (종료·숨김은 빠진다)
//   플레이스    isActive != false
//   장소대여    isActive != false
//
// 지난 파티·끝난 이벤트·숨긴 장소는 **자리를 차지하지 않는다.** 그것들은
// 어차피 14일 뒤 자동 삭제되는 대기열이고, 그 사이에 새 등록이 막히면
// 호스트는 지우는 방법을 찾아 헤매게 된다. 삭제된 것(isDeleted / status
// 'deleted')도 물론 빠진다.
//
// ── 파티는 문서가 아니라 '게시글' 단위다 ─────────────────────────────────
// 날짜를 여러 개 고른 일회성 파티는 **날짜마다 문서 하나**가 만들어지고, 그
// 문서들은 같은 `seriesId`를 공유한다(party_register_screen.dart / createParty).
// 사용자가 만든 것은 파티 하나이므로 seriesId로 묶어 하나로 센다. seriesId가
// 없는 옛 문서는 문서 자신을 한 건으로 본다.
//
// 정기 파티는 애초에 문서 하나(+ recurringSchedule)라 언제나 1이다.
//
// ── 수정은 늘어나지 않는다 ───────────────────────────────────────────────
// 세는 것은 "지금 있는 것"이지 "만든 횟수"가 아니다. 기존 문서를 수정하면
// 개수가 그대로이므로 한도에 걸리지 않는다.
//
// ── 클라이언트와 같은 규칙이어야 한다 ────────────────────────────────────
// 앱 쪽 정본은 party_app/lib/services/registration_limits.dart —
// [RegistrationLimits]. 두 곳의 상한(10)과 "무엇을 세는가"가 어긋나면 앱은
// 통과시키고 서버가 지우는(또는 그 반대) 상태가 된다.
// ══════════════════════════════════════════════════════════════════════════

const { finalPartyEndAt } = require('./partySchedule');

/// 유형별 상한. 네 유형이 같은 값을 쓴다 — 유형마다 다른 숫자를 두면
/// 안내 문구도 유형마다 갈라진다.
const MAX_PER_TYPE = 10;

/// 컬렉션 이름 — 위 표의 키 → 실제 컬렉션.
const COLLECTIONS = {
  party: 'parties',
  promotion: 'placePromotions',
  event: 'events',
  place: 'places',
};

/// 사용자에게 그대로 보여줄 이름.
const LABELS = {
  party: '파티',
  promotion: '이벤트',
  event: '플레이스',
  place: '장소대여',
};

/// 한도에 걸렸을 때의 안내 — 앱·웹·서버가 같은 문장을 쓴다.
function limitMessage(type) {
  const label = LABELS[type] || '등록물';
  return `${label}는 최대 ${MAX_PER_TYPE}개까지 등록할 수 있어요. `
    + '기존 등록물을 삭제하거나 종료한 뒤 다시 시도해주세요.';
}

function toDate(value) {
  if (!value) return null;
  if (typeof value.toDate === 'function') return value.toDate();
  if (value instanceof Date) return value;
  if (typeof value === 'number') return new Date(value);
  return null;
}

/// soft delete — 네 유형이 같은 두 필드를 쓴다.
function isDeleted(data) {
  return data.isDeleted === true || data.status === 'deleted';
}

// ── 유형별 "살아 있는가" 판정 ─────────────────────────────────────────────
//
// 각 판정은 문서 하나만 보는 순수 함수다(테스트가 Firestore 없이 돈다).

/// 파티 문서 하나가 아직 안 끝났는가.
///
/// [finalPartyEndAt]이 null이면 "끝나는 때가 없다"(종료일 없는 무기한 정기
/// 파티)는 뜻이라 살아 있는 것으로 본다 — deleteExpiredParties도 그런 파티는
/// 지우지 않는다. 두 판정이 어긋나면 "세지도 않는데 지워지지도 않는" 파티가
/// 생긴다.
function isPartyDocLive(data, now) {
  if (isDeleted(data)) return false;
  const end = finalPartyEndAt(data);
  if (end == null) return true;
  return end.getTime() >= now.getTime();
}

/// 이벤트(placePromotions) 한 건이 손님에게 보이거나 곧 보일 것인가.
///
/// 앱의 [PlacePromotion.statusAt]과 **같은 규칙**이다:
///   · isVisible == false            → 숨김(호스트가 '종료'를 눌렀다)
///   · isAlways                      → 상시 진행
///   · endAt이 지났으면              → 종료
///   · 그 외(시작 전 포함)           → 살아 있다
/// 종료 판정은 그 날의 끝(23:59:59)까지 본다 — 앱과 같은 하루 단위다.
function isPromotionLive(data, now) {
  if (isDeleted(data)) return false;
  if (data.isVisible === false) return false;
  if (data.isAlways === true) return true;
  const end = toDate(data.endAt);
  if (end == null) return true;
  const dayEnd = new Date(
    end.getFullYear(), end.getMonth(), end.getDate(), 23, 59, 59, 999,
  );
  return dayEnd.getTime() >= now.getTime();
}

/// 이벤트가 **손님에게 보이지 않게 된 시각** — 자동 삭제(14일)를 세는 기준점.
///
/// 그렇게 되는 길은 둘이고, 둘 다 겪었다면 **먼저 일어난 쪽**이 기준이다.
///   · 기간 종료 — endAt이 가리키는 날의 끝(23:59:59)
///   · 호스트가 '종료'를 누름 — hiddenAt
///
/// 아직 보이는 이벤트, 상시 진행(노출 중), 기준점을 모르는 옛 숨김 이벤트는
/// null이다 — **null이면 지우지 않는다.**
///
/// ⚠️ 앱의 [PlacePromotion.goneAt]과 같은 규칙이어야 한다. 앱이 'D-3'이라고
///    써 놓고 서버가 이미 지웠으면 안 된다.
function promotionGoneAt(data, now = new Date()) {
  if (!data) return null;
  if (data.isDeleted === true || data.status === 'deleted') return null;
  const stamps = [];

  const end = toDate(data.endAt);
  if (end && data.isAlways !== true) {
    const dayEnd = new Date(
      end.getFullYear(), end.getMonth(), end.getDate(), 23, 59, 59, 999,
    );
    if (dayEnd.getTime() < now.getTime()) stamps.push(dayEnd.getTime());
  }

  const hidden = toDate(data.hiddenAt);
  if (data.isVisible === false && hidden) stamps.push(hidden.getTime());

  if (stamps.length === 0) return null;
  return new Date(Math.min(...stamps));
}

/// 플레이스·장소대여는 날짜가 없는 상시 등록물이라 **숨김 여부**가 전부다.
/// 값이 없는 옛 문서는 노출 중으로 본다(목록 판정 ListingSources와 같다).
function isPlaceDocLive(data) {
  if (isDeleted(data)) return false;
  return data.isActive !== false;
}

/// 문서 하나가 그 유형의 한도에 자리를 차지하는가.
function isLive(type, data, now) {
  switch (type) {
    case 'party': return isPartyDocLive(data, now);
    case 'promotion': return isPromotionLive(data, now);
    case 'event':
    case 'place': return isPlaceDocLive(data);
    default: return false;
  }
}

/// 문서를 "사용자가 만든 것 하나" 단위로 묶는 열쇠.
///
/// 파티만 seriesId로 묶인다(위 주석). 나머지는 문서 하나가 곧 등록물 하나다.
function groupKeyOf(type, docId, data) {
  if (type !== 'party') return docId;
  const series = typeof data.seriesId === 'string' ? data.seriesId.trim() : '';
  return series.length > 0 ? series : docId;
}

/// 문서 목록에서 **지금 자리를 차지하는 등록물 수**를 센다.
///
/// [excludeKeys]는 판정에서 빼고 셀 열쇠들이다 — 방금 만들어진 문서가 자기
/// 자신 때문에 한도를 넘긴 것처럼 보이지 않게 할 때 쓴다.
function countLiveEntries(type, docs, { now = new Date(), excludeKeys = [] } = {}) {
  const skip = new Set(excludeKeys);
  const keys = new Set();
  for (const doc of docs) {
    const data = doc.data;
    if (!isLive(type, data, now)) continue;
    const key = groupKeyOf(type, doc.id, data);
    if (skip.has(key)) continue;
    keys.add(key);
  }
  return keys.size;
}

// ── Firestore를 실제로 읽는 자리 ─────────────────────────────────────────

/// 이 호스트가 지금 갖고 있는 [type] 등록물 수.
///
/// hostId 하나로만 거르고 나머지는 메모리에서 판정한다 — 상태 축이 유형마다
/// 다르고(종료 시각·isVisible·isActive) 색인을 넷이나 늘릴 만한 쿼리가 아니다.
/// 한 호스트의 문서는 상한이 10 언저리라 읽는 양도 작다.
async function countActive(db, type, hostId, { now = new Date(), excludeKeys = [] } = {}) {
  const collection = COLLECTIONS[type];
  if (!collection || !hostId) return 0;
  const snap = await db.collection(collection).where('hostId', '==', hostId).get();
  const docs = snap.docs.map((d) => ({ id: d.id, data: d.data() }));
  return countLiveEntries(type, docs, { now, excludeKeys });
}

/// 새로 하나 더 만들 수 있는가. (한도에 **닿으면** 더는 못 만든다.)
async function canCreate(db, type, hostId, options = {}) {
  const count = await countActive(db, type, hostId, options);
  return count < MAX_PER_TYPE;
}

module.exports = {
  MAX_PER_TYPE,
  COLLECTIONS,
  LABELS,
  limitMessage,
  isDeleted,
  isPartyDocLive,
  isPromotionLive,
  promotionGoneAt,
  isPlaceDocLive,
  isLive,
  groupKeyOf,
  countLiveEntries,
  countActive,
  canCreate,
};

File: party_app/lib/services/registration_limits.dart
// ─────────────────────────────────────────────────────────────────────────────
// 등록 개수 제한 — **유형별 각각 10개**. 앱에서 이 판정을 하는 유일한 자리.
//
// 한 계정이 동시에 운영할 수 있는 등록물 수를 유형마다 따로 센다. 합산이
// 아니다 — 파티 10 + 이벤트 10 + 플레이스 10 + 장소대여 10까지 된다.
//
// ── 왜 한 곳인가 ─────────────────────────────────────────────────────────────
// 예전에는 파티만, 그것도 **두 곳에서 서로 다른 기준으로** 셌다. 등록 진입
// 화면은 parties 문서를 그대로 셌고(날짜 5개짜리 파티 = 5개), 저장 직전
// 검사는 seriesId로 묶어 셌다(= 1개). 같은 계정이 진입에서는 막히고 저장에서는
// 통과하는 구간이 실제로 있었다. 이제 세는 규칙도, 숫자도, 안내 문구도 여기
// 하나뿐이다.
//
// ── 서버가 마지막 문이다 ─────────────────────────────────────────────────────
// 여기는 **화면을 막는 자리**일 뿐이다. 실제 강제는 서버가 한다 —
// functions/registrationLimits.js(같은 규칙)와 그것을 쓰는
// registrationLimitGuard(네 컬렉션의 onCreate 트리거) · createParty 콜러블.
// 두 쪽의 상한과 "무엇을 세는가"가 어긋나면 앱은 통과시키고 서버가 지우는
// 상태가 된다.
//
// ── 무엇이 자리를 차지하는가 ─────────────────────────────────────────────────
// **지금 살아 있는 것만** 센다.
//   파티      최종 종료 시각이 아직 안 지난 것(무기한 정기 파티 포함)
//   이벤트    진행 중 · 시작 전([PlacePromotion.statusAt]이 공개로 보는 것)
//   플레이스  isActive != false
//   장소대여  isActive != false
//
// 지난 파티·끝난 이벤트·숨긴 장소는 자리를 차지하지 않는다 — 어차피 14일 뒤
// 자동 삭제되는 대기열이고([AutoDeleteRetention]), 그 사이에 새 등록이 막히면
// 호스트는 지울 방법을 찾아 헤매게 된다. 삭제된 것도 물론 빠진다.
//
// 파티만 **게시글 단위**다. 날짜를 여러 개 고른 일회성 파티는 날짜마다 문서가
// 생기고 그 문서들이 `seriesId`를 공유한다 — 사용자가 만든 것은 하나이므로
// 하나로 센다.
//
// 수정은 개수를 늘리지 않는다: 세는 것은 "지금 있는 것"이지 "만든 횟수"가
// 아니다.
// ─────────────────────────────────────────────────────────────────────────────

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';

import 'package:party_app/models/party_schedule.dart';
import 'package:party_app/models/place_promotion.dart';
import 'package:party_app/utils/user_session.dart';

/// 등록물 유형 — 컬렉션과 사람이 부르는 이름이 어긋나 있어서 **둘 다** 적어
/// 둔다(서버 registrationLimits.js의 키와 같은 이름이다).
enum RegistrationKind {
  /// 🎉 파티 — `parties`.
  party(key: 'party', collection: 'parties', label: '파티'),

  /// ✨ 이벤트(매장 이벤트 · 공간 이벤트) — `placePromotions`.
  promotion(key: 'promotion', collection: 'placePromotions', label: '이벤트'),

  /// 🍷 플레이스(술집·바·카페) — `events`. **컬렉션 이름에 속지 말 것.**
  placeListing(key: 'event', collection: 'events', label: '플레이스'),

  /// 🏠 장소대여(공간대여·숙박) — `places`.
  rentalListing(key: 'place', collection: 'places', label: '장소대여');

  const RegistrationKind({
    required this.key,
    required this.collection,
    required this.label,
  });

  /// 서버와 맞춘 유형 키.
  final String key;

  /// 실제 Firestore 컬렉션.
  final String collection;

  /// 안내 문구에 쓰는 이름.
  final String label;
}

/// 목록에서 셀 문서 한 건 — Firestore 없이 판정을 테스트하려고 둔 최소 형태.
class RegistrationDoc {
  const RegistrationDoc(this.id, this.data);

  final String id;
  final Map<String, dynamic> data;
}

class RegistrationLimits {
  RegistrationLimits._();

  /// 유형별 상한. 네 유형이 같은 값을 쓴다 — 유형마다 다른 숫자를 두면
  /// 안내 문구도 유형마다 갈라진다.
  ///
  /// ⚠️ functions/registrationLimits.js의 MAX_PER_TYPE과 **같아야 한다.**
  static const int maxPerType = 10;

  /// 한도에 걸렸을 때 보여줄 안내. 서버 limitMessage와 같은 뜻이다.
  static String limitMessage(RegistrationKind kind) =>
      '${kind.label}는 최대 $maxPerType개까지 등록할 수 있어요.\n'
      '기존 등록물을 삭제하거나 종료한 뒤 다시 시도해주세요.';

  // ── 판정(순수) ─────────────────────────────────────────────────────────

  /// soft delete — 네 유형이 같은 두 필드를 쓴다.
  static bool _isDeleted(Map<String, dynamic> d) =>
      d['isDeleted'] == true || d['status'] == 'deleted';

  /// 문서 하나가 지금 자리를 차지하는가.
  static bool isLive(
    RegistrationKind kind,
    RegistrationDoc doc, {
    DateTime? now,
  }) {
    final data = doc.data;
    if (_isDeleted(data)) return false;
    final at = now ?? DateTime.now();
    switch (kind) {
      case RegistrationKind.party:
        // 끝나는 때가 없는 무기한 정기 파티(null)는 살아 있다 — 서버의 자동
        // 삭제도 그런 파티는 건드리지 않는다. 두 판정이 어긋나면 "세지도
        // 않는데 지워지지도 않는" 파티가 생긴다.
        final end = PartySchedule.finalEndAt(data);
        return end == null || !end.isBefore(at);
      case RegistrationKind.promotion:
        // 종료·숨김만 빠진다(진행 중·시작 전은 남는다) — 상세·목록이 쓰는
        // 판정 하나를 그대로 쓴다.
        return PlacePromotion.fromMap(doc.id, data).statusAt(at).isPublic;
      case RegistrationKind.placeListing:
      case RegistrationKind.rentalListing:
        // 날짜가 없는 상시 등록물이라 숨김 여부가 전부다. 값이 없는 옛 문서는
        // 노출 중으로 본다(목록 판정 ListingSources와 같다).
        return data['isActive'] != false;
    }
  }

  /// 문서를 "사용자가 만든 것 하나" 단위로 묶는 열쇠 — 파티만 seriesId다.
  static String groupKeyOf(RegistrationKind kind, RegistrationDoc doc) {
    if (kind != RegistrationKind.party) return doc.id;
    final series = (doc.data['seriesId'] as String? ?? '').trim();
    return series.isNotEmpty ? series : doc.id;
  }

  /// 목록에서 지금 자리를 차지하는 등록물 수.
  ///
  /// [excludeKeys]에 든 열쇠는 빼고 센다 — "이것 말고 몇 개가 있나"를 물어야
  /// 하는 자리(수정 중인 문서 등)를 위해 둔다.
  static int countLive(
    RegistrationKind kind,
    Iterable<RegistrationDoc> docs, {
    DateTime? now,
    Set<String> excludeKeys = const {},
  }) {
    final keys = <String>{};
    for (final doc in docs) {
      if (!isLive(kind, doc, now: now)) continue;
      final key = groupKeyOf(kind, doc);
      if (excludeKeys.contains(key)) continue;
      keys.add(key);
    }
    return keys.length;
  }

  // ── Firestore ──────────────────────────────────────────────────────────

  /// 테스트 주입점 — 이 프로젝트의 Dart 테스트에는 Firestore 페이크가 없다
  /// ([PlaceCreateEligibility]와 같은 방식). 운영 코드는 null이라 실물
  /// Firestore를 그대로 쓴다.
  static Future<List<RegistrationDoc>> Function(RegistrationKind kind)? _source;

  @visibleForTesting
  static void debugSetSource(
    Future<List<RegistrationDoc>> Function(RegistrationKind kind)? source,
  ) => _source = source;

  @visibleForTesting
  static void debugResetSource() => _source = null;

  static Future<List<RegistrationDoc>> _fetch(RegistrationKind kind) async {
    final override = _source;
    if (override != null) return override(kind);
    final uid = UserSession.userId;
    if (uid.isEmpty) return const [];
    // hostId 하나로만 거르고 상태는 메모리에서 본다 — 상태 축이 유형마다
    // 다르고(종료 시각·isVisible·isActive), 한 계정의 문서는 많아야 수십 건이다.
    // 캐시를 우회한다: 방금 지운 등록물이 캐시에 남아 한도를 잘못 채우면
    // 사용자는 "지웠는데도 못 만드는" 상태를 만난다.
    final snap = await FirebaseFirestore.instance
        .collection(kind.collection)
        .where('hostId', isEqualTo: uid)
        .get(const GetOptions(source: Source.server));
    return [for (final d in snap.docs) RegistrationDoc(d.id, d.data())];
  }

  /// 지금 이 계정이 갖고 있는 [kind] 등록물 수.
  static Future<int> countActive(
    RegistrationKind kind, {
    DateTime? now,
    Set<String> excludeKeys = const {},
  }) async {
    final docs = await _fetch(kind);
    return countLive(kind, docs, now: now, excludeKeys: excludeKeys);
  }

  /// 하나 더 만들 수 있는가. 조회에 실패하면 **막지 않는다** — 개수 제한은
  /// 안전장치지 관문이 아니고(진짜 관문은 사업자 인증이다), 서버가 초과분을
  /// 다시 잡아낸다. 여기서 fail-closed로 막으면 네트워크가 흔들릴 때마다
  /// 등록 자체가 안 되는 것처럼 보인다.
  static Future<bool> canCreate(RegistrationKind kind, {DateTime? now}) async {
    try {
      return await countActive(kind, now: now) < maxPerType;
    } catch (_) {
      return true;
    }
  }

  // ── 화면에서 쓰는 관문 ─────────────────────────────────────────────────

  /// 등록을 시작해도 되는지 확인하고, 꽉 찼으면 안내를 띄운 뒤 false.
  ///
  /// 등록 폼을 여는 모든 길이 이 함수를 지나야 한다 — 한 곳이라도 빠지면
  /// 그 길이 그대로 우회로가 된다(사업자 인증 관문과 같은 규칙).
  static Future<bool> ensure(
    BuildContext context,
    RegistrationKind kind, {
    DateTime? now,
  }) async {
    if (await canCreate(kind, now: now)) return true;
    if (!context.mounted) return false;
    await showDialog<void>(
      context: context,
      builder: (_) => AlertDialog(
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
        title: const Text(
          '등록 개수를 다 채웠어요',
          style: TextStyle(
            fontFamily: 'SeoulHangang',
            fontWeight: FontWeight.w500,
            shadows: [
              Shadow(color: Colors.black87, offset: Offset(0.3, 0)),
              Shadow(color: Colors.black87, offset: Offset(-0.3, 0)),
              Shadow(color: Colors.black87, offset: Offset(0, 0.3)),
              Shadow(color: Colors.black87, offset: Offset(0, -0.3)),
            ],
          ),
        ),
        content: Text(
          limitMessage(kind),
          style: const TextStyle(height: 1.5),
        ),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: const Text(
              '확인',
              style: TextStyle(color: Color(0xFFFF6FA0)),
            ),
          ),
        ],
      ),
    );
    return false;
  }
}
